client.go

```diff
@@ -87,11 +198,269 @@ func (c *Client) SendSTUNRequest(serverIP net.IP, serverPort int) (net.Addr, err
+		return nil, err
+	}
+	c.realm = append([]byte(nil), c.realm...)
+	c.integrity = stun.NewLongTermIntegrity(
+		c.username.String(), c.realm.String(), c.password,
+	)
+	// Trying to authorize.
+	msg, err = stun.Build(
+		stun.TransactionID,
+		stun.NewType(stun.MethodAllocate, stun.ClassRequest),
+		turn.RequestedTransportUDP,
+		&c.username,
+		&c.realm,
+		&nonce,
+		&c.integrity,
+		stun.Fingerprint,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	trRes, err = c.PerformTransaction(msg, c.turnServ, false)
+	if err != nil {
+		return nil, err
+	}
+	res = trRes.Msg
+
+	if res.Type.Class == stun.ClassErrorResponse {
+		var code stun.ErrorCodeAttribute
+		if err = code.GetFrom(res); err == nil {
+			return nil, fmt.Errorf("%s (error %s)", res.Type, code)
+		}
+		return nil, fmt.Errorf("%s", res.Type)
+	}
+
+	// Getting relayed addresses from response.
+	var relayed turn.RelayedAddress
+	if err := relayed.GetFrom(res); err != nil {
+		return nil, err
+	}
+
+	// Getting lifetime from response
+	var lifetime turn.Lifetime
+	if err := lifetime.GetFrom(res); err != nil {
+		return nil, err
+	}
+
+	relayedConn = client.NewUDPConn(&client.UDPConnConfig{
+		Observer:  c,
+		Relayed:   relayed,
+		Integrity: c.integrity,
+		Nonce:     nonce,
+		Lifetime:  lifetime.Duration,
+		Log:       c.log,
+	})
+
+	c.mutex.Lock()
+	c.relayedConn = relayedConn
+	c.mutex.Unlock()
+
+	return relayedConn, nil
+}
+
+// PerformTransaction performs STUN transaction
+func (c *Client) PerformTransaction(msg *stun.Message, to net.Addr, dontWait bool) (client.TrResult, error) {
+	trKey := b64.StdEncoding.EncodeToString(msg.TransactionID[:])
+	tr := client.NewTransaction(&client.TransactionConfig{
+		Key:      trKey,
+		Raw:      msg.Raw,
+		To:       to,
+		Interval: c.rto,
+	})
+
+	c.trMap.Insert(trKey, tr)
+
+	c.log.Tracef("start %s transaction %s to %s", msg.Type, trKey, tr.To.String())
+	_, err := c.conn.WriteTo(msg.Raw, to)
+	if err != nil {
+		return client.TrResult{}, err
+	}
+
+	tr.StartRtxTimer(c.onRtxTimeout)
+
+	// If dontWait is true, get the transaction going and return immediately
+	if dontWait {
+		return client.TrResult{}, nil
+	}
+
+	res := tr.WaitForResult()
+	if res.Err != nil {
+		return res, res.Err
+	}
+	return res, nil
+}
+
+// HandleInbound handles data received.
+// This method handles packets received only from the turn server address.
+// If the source (from) address does not match, it would return (false, nil)
+// to indicate the caller that the packet was not handled.
+func (c *Client) HandleInbound(data []byte, from net.Addr) (bool, error) {
+	c.log.Debug("HandleInbound: in")
+	defer c.log.Debug("HandleInbound: out")
+	var err error
+
+	switch {
+	case stun.IsMessage(data):
+		err = c.handleSTUNMessage(data, from)
+	case turn.IsChannelData(data):
+		err = c.handleChannelData(data)
+	default:
+		// assume, this is an application data
+		c.log.Tracef("non-STUN/TURN packect, unhandled")
+		return false, nil // unhandled
+	}
+
+	if err != nil {
```

Comment:
I would always `return true, err` (if err is nil we are returning the same thing)

Reply:
ah ok, will do

Reply:
Done